src/shogun/structure/GEMPLP.cpp

```diff
@@ -0,0 +1,368 @@
+				assignment[m_all_separators[s][0]] = max_at;
+		}
+		
+		// get the value of the decoded solution
+		float64_t int_val = 0;
+	
+		// iterates over factors
+		for (int32_t c = 0; c < m_factors->get_num_elements(); c++)
+		{
+			CFactor* factor = dynamic_cast<CFactor*>(m_factors->get_element(c));
+			SGVector<int32_t> vars = factor->get_variables();
+			SGVector<int32_t> var_assignment(vars.size());
+			
+			for (int32_t i = 0; i < vars.size(); i++)
+				var_assignment[i] = assignment[vars[i]];
+
+			// add value from current factor
+			int_val += m_theta_cluster[c].get_value(var_assignment);
+
+			SG_UNREF(factor);
+		}
+		
+		float64_t obj_del = last_obj - obj;
+		float64_t int_gap = obj - int_val;
+
+		SG_SDEBUG("Iter= %d Objective=%f ObjBest=%f ObjDel=%f Gap=%f \n", (it + 1), obj, int_val, obj_del, int_gap);
+
+		if (obj_del < m_param.m_obj_del_thr && it > 16)
+			break;
+		
+		if (int_gap < m_param.m_int_gap_thr)
+			break;
+
+		last_obj = obj;
+	}
+	
+	float64_t energy = m_fg->evaluate_energy(assignment);
+	SG_DEBUG("fg.evaluate_energy(assignment) = %f\n", energy);
+
+	return energy;
+}
+
+void CGEMPLP::update_messages(int32_t id_clique)
+{
+	CFactor* factor = dynamic_cast<CFactor*>(m_factors->get_element(id_clique));
+	SGVector<int32_t> vars = factor->get_variables();
+	SGVector<int32_t> cards = factor->get_cardinalities();
+	SGNDArray<float64_t> lam_sum(cards);
+
+	if (m_theta_cluster[id_clique].len_array == 0)
+		lam_sum.set_const(0);
+	else
+		lam_sum = m_theta_cluster[id_clique].clone();
+
+	int32_t num_separators = m_clique_separators[id_clique].size();
+	vector<SGNDArray<float64_t> > lam_minus; // substract message: \lambda_s^{-c}(x_s)
+	// \sum_{\hat{s}} \lambda_{\hat{s}}^{-c}(x_{\hat{s}}) + \theta_c(x_c)
+	for (int32_t s = 0; s < num_separators; s++)
+	{
+		int32_t id_separator = m_clique_separators[id_clique][s];
+		SGNDArray<float64_t> tmp = m_msgs_into_separators[id_separator].clone();
+		tmp -= m_msgs_from_cluster[id_clique][s];
+		
+		lam_minus.push_back(tmp);
+		
+		if (vars.size() == (int32_t)m_clique_inds_separators[id_clique][s].size())
+			lam_sum += tmp;
+		else
+		{
+			SGNDArray<float64_t> tmp_expand(lam_sum.get_dimensions());
+			tmp.expand(tmp_expand, m_clique_inds_separators[id_clique][s]);
+			lam_sum += tmp_expand;
+		}
+
+		// take out the old incoming message: \lambda_{c \to s}(x_s)
+		m_msgs_into_separators[id_separator] -= m_msgs_from_cluster[id_clique][s];
+	}
+	
+	for (int32_t s = 0; s < num_separators; s++)
+	{
+		// maximazation: \max_{x_c} \sum_{\hat{s}} \lambda_{\hat{s}}^{-c}(x_{\hat{s}}) + \theta_c(x_c)
+		SGNDArray<float64_t> lam_max(lam_minus[s].get_dimensions());
+		max_in_subdimension(lam_sum, m_clique_inds_separators[id_clique][s], lam_max);
+		int32_t id_separator = m_clique_separators[id_clique][s];
+		// weighted sum
+		lam_max *= 1.0/num_separators;
+		m_msgs_from_cluster[id_clique][s] = lam_max.clone();
+		m_msgs_from_cluster[id_clique][s] -= lam_minus[s];
```

Comment:
Yes, you are right. In fact, I am confused by this line of code as well. The original implementation does not  use `lam_minus *= 1.0/num_separators;`. I will check it later.